File: Types/OpenTokBroadcast.fs
namespace OpenTokFs.Types

open System

/// <summary>
/// An RTMP destination for an OpenTok broadcast.
/// </summary>
[<AllowNullLiteral>]
type OpenTokRtmpStream() =
    member val Id: string = "" with get, set
    member val ServerUrl: string = "" with get, set
    member val StreamName: string = "" with get, set
    member val Status: string = "" with get, set

/// <summary>
/// Streaming endpoints for an OpenTok broadcast.
/// </summary>
[<AllowNullLiteral>]
type OpenTokBroadcastStreams() =
    member val Hls: string = "" with get, set
    member val Rtmp: OpenTokRtmpStream[] = Array.empty with get, set

/// <summary>
/// A running OpenTok broadcast.
/// </summary>
[<AllowNullLiteral>]
type OpenTokBroadcast() =
    member val Id: string = "" with get, set
    member val SessionId: string = "" with get, set
    member val ProjectId: int = 0 with get, set
    member val CreatedAt: int64 = 0L with get, set
    member val BroadcastUrls: OpenTokBroadcastStreams = new OpenTokBroadcastStreams() with get, set
    member val UpdatedAt: int64 = 0L with get, set
    member val Status: string = "" with get, set
    member val MaxDuration: int = 0 with get, set
    member val Resolution: string = "" with get, set

    member this.GetCreationTime() = DateTimeOffset.FromUnixTimeMilliseconds this.CreatedAt
    member this.GetUpdatedTime() = DateTimeOffset.FromUnixTimeMilliseconds this.UpdatedAt
    member this.GetMaxDuration() = this.MaxDuration |> float |> TimeSpan.FromSeconds

File: OpenTokFs.Json/ResponseTypes/OpenTokBroadcast.cs
using System;
using System.Collections.Generic;

namespace OpenTokFs.Json.ResponseTypes {
    public class OpenTokRtmpStream {
        public string Id { get; set; }
        public string ServerUrl { get; set; }
        public string StreamName { get; set; }
        public string Status { get; set; }

        public override string ToString() => $"{ServerUrl}/{StreamName} ({Id ?? "no ID"}) ({Status})";
    }

    public class OpenTokBroadcastStreams {
        public string Hls { get; set; }
        public IEnumerable<OpenTokRtmpStream> Rtmp { get; set; }

        public override string ToString() {
            return $"HLS: {Hls ?? "none"}; RTMP: [{string.Join(", ", Rtmp)}]";
        }
    }

    public class OpenTokBroadcast {
        public string Id { get; set; }
        public string SessionId { get; set; }
        public int ProjectId { get; set; }
        public long CreatedAt { get; set; }
        public OpenTokBroadcastStreams BroadcastUrls { get; set; }
        public long UpdatedAt { get; set; }
        public string Status { get; set; }
        public string Resolution { get; set; }

        public DateTimeOffset GetCreationTime() => DateTimeOffset.FromUnixTimeMilliseconds(CreatedAt);
        public DateTimeOffset GetUpdatedTime() => DateTimeOffset.FromUnixTimeMilliseconds(UpdatedAt);

        public override string ToString() => $"{Id} ({Status})";
    }
}
